src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToNewComponentCodeActionProvider.cs

```diff
@@ -47,51 +51,169 @@ public Task<ImmutableArray<RazorVSInternalCodeAction>> ProvideAsync(RazorCodeAct
+        {
+            Action = LanguageServerConstants.CodeActions.ExtractToNewComponentAction,
+            Language = LanguageServerConstants.CodeActions.Languages.Razor,
+            Data = actionParams,
+        };
+
+        var codeAction = RazorCodeActionFactory.CreateExtractToNewComponent(resolutionParams);
+        return Task.FromResult<ImmutableArray<RazorVSInternalCodeAction>>([codeAction]);
+    }
+
+    private static (MarkupElementSyntax? Start, MarkupElementSyntax? End) GetStartAndEndElements(RazorCodeActionContext context, RazorSyntaxTree syntaxTree, ILogger logger)
+    {
         var owner = syntaxTree.Root.FindInnermostNode(context.Location.AbsoluteIndex, includeWhitespace: true);
         if (owner is null)
         {
-            _logger.LogWarning($"Owner should never be null.");
-            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+            logger.LogWarning($"Owner should never be null.");
+            return (null, null);
+        }
+
+        var startElementNode = owner.FirstAncestorOrSelf<MarkupElementSyntax>();
+        if (startElementNode is null || IsInsideProperHtmlContent(context, startElementNode))
+        {
+            return (null, null);
         }
 
-        var componentNode = owner.FirstAncestorOrSelf<MarkupElementSyntax>();
+        var endElementNode = GetEndElementNode(context, syntaxTree);
+        return (startElementNode, endElementNode);
+    }
 
-        // Make sure we've found tag
-        if (componentNode is null)
+    private static bool IsInsideProperHtmlContent(RazorCodeActionContext context, MarkupElementSyntax startElementNode)
+    {
+        // If the provider executes before the user/completion inserts an end tag, the below return fails
+        if (startElementNode.EndTag.IsMissing)
         {
-            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+            return true;
         }
 
-        // Do not provide code action if the cursor is inside proper html content (i.e. page text)
-        if (context.Location.AbsoluteIndex > componentNode.StartTag.Span.End &&
-            context.Location.AbsoluteIndex < componentNode.EndTag.SpanStart)
+        return context.Location.AbsoluteIndex > startElementNode.StartTag.Span.End &&
+               context.Location.AbsoluteIndex < startElementNode.EndTag.SpanStart;
+    }
+
+    private static MarkupElementSyntax? GetEndElementNode(RazorCodeActionContext context, RazorSyntaxTree syntaxTree)
+    {
+        var selectionStart = context.Request.Range.Start;
+        var selectionEnd = context.Request.Range.End;
+        if (selectionStart == selectionEnd)
         {
-            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+            return null;
         }
 
-        if (!TryGetNamespace(context.CodeDocument, out var @namespace))
+        var endLocation = GetEndLocation(selectionEnd, context.CodeDocument);
```

Comment:
I don't think we need `GetEndLocation` to exist, this can just be `var endAbsoluteIndex = context.SourceText.GetRequiredAbsoluteIndex(selectionEnd)`.